cli/debug: debugging improvements

andreimatei opened the pull request:
See individual commits.

Release note: None

andreimatei committed:
cli: print key range hex in "debug raft-log"

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/35528)
<!-- Reviewable:end -->

tbg commented on pkg/cli/debug.go:
```diff
@@ -449,6 +449,7 @@ func runDebugRaftLog(cmd *cobra.Command, args []string) error {
 
 	start := engine.MakeMVCCMetadataKey(keys.RaftLogPrefix(rangeID))
 	end := engine.MakeMVCCMetadataKey(keys.RaftLogPrefix(rangeID).PrefixEnd())
+	fmt.Printf("Printing keys %s -> %s (%#x - %#x)\n", start, end, start, end)
```
Doesn't do what you want -- you need to encode this MVCCMetadata key down to `[]byte` and then print it as hex.

tbg replied:
i.e. you want the opposite of https://github.com/cockroachdb/cockroach/blob/e46eaeeb01f31bfbcbf05b623361ea2e047ddb6c/pkg/storage/engine/enginepb/decode.go#L52

Not sure if we have it on the Go side.

tbg replied:
nvm, here it is: https://github.com/cockroachdb/cockroach/blob/50bba75d75e189f6d0496d2a940ee09c1cf82504/pkg/storage/engine/batch.go#L305

andreimatei reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35528)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @tbg)

___
*[pkg/cli/debug.go, line 452 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35528#-L_P8u8RieZmFGGF4ksf-r1-452:-L_U5N1-0iEdXEmzzKdw:b54lxdw) ([raw file](https://github.com/cockroachdb/cockroach/blob/52bc7f5ef45ab56dce12f9d333604c51bcf6a62f/pkg/cli/debug.go#L452)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

nvm, here it is: https://github.com/cockroachdb/cockroach/blob/50bba75d75e189f6d0496d2a940ee09c1cf82504/pkg/storage/engine/batch.go#L305
</blockquote></details>

done, thanks.

Btw, how did you come up with the key encoding for the `RangeAppliedState` queries you've done recently?


<!-- Sent from Reviewable.io -->

andreimatei committed:
cli/debug: remove unused param

Release note: None

andreimatei committed:
cli/debug: teach debug raft-log to decode write batches

For each command, looks like:
write batch: 0.000000000,0 /Local/Range/Table/56/1/"\x04\x02z\xad\xbe\x14Dɤ\xc5\xf8\xf1w\x153F"/PrefixEnd/RangeDescriptor (0x016b12c0891204027aadbe1444c9a4c5f8f17715334600ff0200017264736300): 1551793064.152233192,5 {Txn:id:51199340-99d5-4493-a12c-ade061f6f92e key:"\001k\022\300\211\022\004\002z\255\276\024D\311\244\305\370\361w\0253F\000\377\002\000\001rdsc" timestamp:<wall_time:1551793064152233192 logical:5 > priority:118000 sequence:1  Timestamp:1551793064.152233192,5 Deleted:false KeyBytes:12 ValBytes:82 RawBytes:[] IntentHistory:[] MergeTimestamp:<nil> XXX_NoUnkeyedLiteral:{} XXX_sizecache:0}
1551793064.152233192,5 /Local/Range/Table/56/1/"\x04\x02z\xad\xbe\x14Dɤ\xc5\xf8\xf1w\x153F"/PrefixEnd/RangeDescriptor (0x016b12c0891204027aadbe1444c9a4c5f8f17715334600ff020001726473630015891391f1b574e8000000050d): [/Table/56/1/"\x04\x02z\xad\xbe\x14Dɤ\xc5\xf8\xf1w\x153F"/PrefixEnd, /Table/56/1/"\x04W\"e\xf2\xecIZ\xa0\x88\x0e!9~K`")
        Raw:r336:/Table/56/1/"\x04{\x02z\xad\xbe\x14Dɤ\xc5\xf8\xf1w\x153F"/PrefixEnd-W\"e\xf2\xecIZ\xa0\x88\x0e!9~K`"} [(n4,s4):4, (n2,s2):2, (n3,s3):3, next=5, gen=2]

Release note: None

andreimatei commented:
Pushed some more commits. PTAL.

tbg reviewed (commented):
> Reviewed 1 of 2 files at r3, 1 of 1 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35528)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei)

___
*[pkg/cli/debug.go, line 452 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/35528#-L_P8u8RieZmFGGF4ksf-r1-452:-L_qn6pv-331B2JAXCu7:bfc5qxa) ([raw file](https://github.com/cockroachdb/cockroach/blob/52bc7f5ef45ab56dce12f9d333604c51bcf6a62f/pkg/cli/debug.go#L452)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

done, thanks.

Btw, how did you come up with the key encoding for the `RangeAppliedState` queries you've done recently?
</blockquote></details>

In my case it was printed by the consistency checker, so I got it for free.
___
*[pkg/cli/debug/print.go, line 95 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/35528#-L_qnrJX0xu4clhGx5sP:-L_qnrJX0xu4clhGx5sQ:b-555knp) ([raw file](https://github.com/cockroachdb/cockroach/blob/85bf2b5ac1207ba9d4be2fd564f559d01767346e/pkg/cli/debug/print.go#L95)):*
> ```Go
> }
> 
> func decodeWriteBatch(writeBatch *storagepb.WriteBatch) (string, error) {
> ```

This is great, but I'll have to ask you to make it a little better because it's probably the last time we're touching it and we'll want it to work correctly. Applying the writebatch to an engine and then reading the engine hides stuff in the writebatch. For example, a writebatch with a deletion in it looks the same as an empty writebatch, simply because the deletion has no effect on an empty underlying engine. I think you want to use this here instead:

https://github.com/cockroachdb/cockroach/blob/50bba75d75e189f6d0496d2a940ee09c1cf82504/pkg/storage/engine/batch.go#L421-L430

The API is straightforward.
___
*[pkg/cli/debug/print.go, line 105 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/35528#-L_qnGjq9FhOD8CRpWGf:-L_qnGjrEHMLJv4GuIKp:bybiqjv) ([raw file](https://github.com/cockroachdb/cockroach/blob/85bf2b5ac1207ba9d4be2fd564f559d01767346e/pkg/cli/debug/print.go#L105)):*
> ```Go
> 		return "", err
> 	}
> 	nilKey := roachpb.Key([]byte{0})
> ```

nit: not really a nil key.


<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35528)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @tbg)

___
*[pkg/cli/debug/print.go, line 95 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/35528#-L_qnrJX0xu4clhGx5sP:-La8QgyE-mHyaHMaVmS_:b-t6wq29) ([raw file](https://github.com/cockroachdb/cockroach/blob/85bf2b5ac1207ba9d4be2fd564f559d01767346e/pkg/cli/debug/print.go#L95)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

This is great, but I'll have to ask you to make it a little better because it's probably the last time we're touching it and we'll want it to work correctly. Applying the writebatch to an engine and then reading the engine hides stuff in the writebatch. For example, a writebatch with a deletion in it looks the same as an empty writebatch, simply because the deletion has no effect on an empty underlying engine. I think you want to use this here instead:

https://github.com/cockroachdb/cockroach/blob/50bba75d75e189f6d0496d2a940ee09c1cf82504/pkg/storage/engine/batch.go#L421-L430

The API is straightforward.
</blockquote></details>

done
respect to @danhhz for writing that thing


<!-- Sent from Reviewable.io -->

tbg reviewed (approved):
> Reviewed 1 of 1 files at r1, 2 of 2 files at r5, 2 of 2 files at r6, 1 of 1 files at r7.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35528)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei and @tbg)

___
*[pkg/cli/debug/print.go, line 42 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/35528#-LaGLIbU1Kyo8wKqsyhG:-LaGLIbV1-AxNI0cJDRM:b-lxpu85) ([raw file](https://github.com/cockroachdb/cockroach/blob/4245a8b0ebace79972a7c1a4453bd0668a6e8b9f/pkg/cli/debug/print.go#L42)):*
> ```Go
> }
> 
> // SprintKey prety-prings the specified MVCCKey.
> ```

pretty
___
*[pkg/cli/debug/print.go, line 113 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/35528#-LaGLOPO78dX_sWpsvvz:-LaGLOPO78dX_sWpsvw-:bh31dc3) ([raw file](https://github.com/cockroachdb/cockroach/blob/4245a8b0ebace79972a7c1a4453bd0668a6e8b9f/pkg/cli/debug/print.go#L113)):*
> ```Go
> 		switch r.BatchType() {
> 		case engine.BatchTypeDeletion:
> 			mvccKey, err := r.MVCCKey()
> ```

Isn't it silly that you're doing this in every block?


<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35528#-:-LaGfqDl9GCOpyj8qtgN:b-ydi9ay)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @tbg)

___
*[pkg/cli/debug/print.go, line 42 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/35528#-LaGLIbU1Kyo8wKqsyhG:-LaGeCDX3lFOLxumjK_c:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4245a8b0ebace79972a7c1a4453bd0668a6e8b9f/pkg/cli/debug/print.go#L42)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

pretty
</blockquote></details>

Done.
___
*[pkg/cli/debug/print.go, line 113 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/35528#-LaGLOPO78dX_sWpsvvz:-LaGeghS60Cm6Sl5LBWg:b-s7w3c9) ([raw file](https://github.com/cockroachdb/cockroach/blob/4245a8b0ebace79972a7c1a4453bd0668a6e8b9f/pkg/cli/debug/print.go#L113)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Isn't it silly that you're doing this in every block?
</blockquote></details>

it seems silly but I'm not sure how best to deal with the `default` case below. Anyway, whatevs, I'll leave it.


<!-- Sent from Reviewable.io -->

tbg commented:
@andreimatei could you backport? Will be good to have this around.

andreimatei commented:
backport is #35901

On Mon, Mar 18, 2019 at 3:12 PM Tobias Grieger <notifications@github.com>
wrote:

> @andreimatei <https://github.com/andreimatei> could you backport? Will be
> good to have this around.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/35528#issuecomment-474060755>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAXBcfvieYPHTd6kAQjju5NFKfEX8Nf5ks5vX-UdgaJpZM4bkMR8>
> .
>
